File: app/api/user/user_service.py
from app.services import db
from app.models.User_model import User
from app.services import logger
from bson.objectid import ObjectId
from typing import List


class UserService:

    collection_name = "user"

    @staticmethod
    async def query() -> List[dict[str, str]]:
        collection = UserService._get_collection()
        cursor = collection.find({})
        users = []
        for document in cursor:
            item = User(**document).to_dict()
            users.append(item)

        return users

    @staticmethod
    async def get_by_username(username: str) -> User | None:
        try:
            collection = UserService._get_collection()
            user = collection.find_one({"username": username})
            return User(**user) if user else None
        except:
            raise ValueError("No match found")

    @staticmethod
    async def get_by_id(id: str) -> User | None:
        try:
            collection = UserService._get_collection()
            user = collection.find_one({"_id": ObjectId(id)})
            return User(**user) if user else None
        except:
            raise ValueError("No match found")

    @staticmethod
    async def add_user(credentials: dict[str, str]) -> User:
        collection = UserService._get_collection()
        collection.insert_one(credentials)
        return User(**credentials)

    @staticmethod
    async def remove(id: str) -> str:
        try:
            collection = UserService._get_collection()
            collection.delete_one({"_id": ObjectId(id)})
            return id
        except:
            raise ValueError("Invalid id")

    @staticmethod
    def _get_collection():
        try:
            collection = db.get_collection(UserService.collection_name)
            return collection
        except Exception as e:
            logger.error(f"error in user service: {e}")

File: app/api/user/user.service.go
package user

import (
	"context"

	"github.com/gDenisLit/item-server-go/cmd/dtos"
	"github.com/gDenisLit/item-server-go/cmd/models"
	"github.com/gDenisLit/item-server-go/cmd/services"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"
)

const collName string = "user"

func Query() ([]models.User, error) {

	collection, err := services.GetDBColletion(collName)
	if err != nil {
		return nil, err
	}

	cursor, err := collection.Find(context.TODO(), bson.M{})
	if err != nil {
		return nil, err
	}

	var users []models.User
	if err = cursor.All(context.TODO(), &users); err != nil {
		return nil, err
	}
	return users, nil
}

func GetById(userId string) (*models.User, error) {
	collection, err := services.GetDBColletion(collName)
	if err != nil {
		return nil, err
	}

	objectId, err := primitive.ObjectIDFromHex(userId)
	if err != nil {
		return nil, err
	}

	user := &models.User{}
	err = collection.FindOne(context.TODO(), bson.M{"_id": objectId}).Decode(user)
	if err != nil {
		return nil, err
	}
	return user, nil
}

func GetByUsername(username string) (*models.User, error) {
	collection, err := services.GetDBColletion(collName)
	if err != nil {
		return nil, err
	}

	user := &models.User{}
	err = collection.FindOne(context.TODO(), bson.M{"username": username}).Decode(user)
	if err != nil {
		return nil, err
	}
	return user, nil
}

func Add(user *dtos.SignupDTO) (*models.User, error) {
	collection, err := services.GetDBColletion(collName)
	if err != nil {
		return nil, err
	}

	res, err := collection.InsertOne(context.TODO(), user)
	if err != nil {
		return nil, err
	}

	objectId := res.InsertedID.(primitive.ObjectID)
	savedUser := &models.User{
		ID:       objectId,
		Username: user.Username,
		Password: user.Password,
		Fullname: user.Fullname,
		ImgUrl:   user.ImgUrl,
		IsAdmin:  user.IsAdmin,
	}
	return savedUser, nil
}
